File: Two_Pointers/3SUM/solution.py
class Solution(object):
    def threeSum(self, nums):
        n = len(nums) 
        res = [] 
        nums.sort()
        i = 0 
        while i < n :  
            if nums[i] > 0:
                break
            target = -nums[i]
            l = i + 1
            r = n - 1
            while l < r:
                curr = nums[l] + nums[r]
                if curr < target:
                    l += 1
                elif curr > target:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    l += 1
                    r -= 1
            while i + 1 < n and nums[i] == nums[i + 1]:
                i += 1
            i += 1 
            
        return res

File: Two_Pointers/3SUM/solution.cpp
class Solution {
    public:
        vector<vector<int>> threeSum(vector<int>& nums) {
            vector<vector<int>> res ; 
            int n = nums.size() ; 
            sort( nums.begin() , nums.end() ) ; 
            for( int i=0 ; i<n ; ++i ){
                if( nums[i] > 0 ) break ; 
                int l = i+1 ; 
                int r = n-1 ; 
                int target = -nums[i] ; 
                while( l < r ){
                    int sum = nums[l] + nums[r] ; 
                    if( sum < target ){
                        l++ ; 
                    }else if( sum > target ){
                        r-- ; 
                    }else {
                        vector<int> sol(3) ; 
                        sol[0] = nums[i] ;
                        sol[1] = nums[l] ;
                        sol[2] = nums[r] ;
                        res.push_back( sol ) ; 
                        while( l+1<r && nums[l+1] == nums[l] ) l++ ; 
                        l++ ; 
                        r-- ; 
                    }
                }
                while( i+1<n && nums[i+1] == nums[i] ) ++i ; 
            }
            return res ; 
        }
    };